Sources/PackageRegistry/RegistryDownloadsManager.swift

```diff
@@ -0,0 +1,310 @@
+                    completion: completion
+                )
+            }
+        } else {
+            // record the pending lookup
+            assert(self.pendingLookups[package] == nil)
+            let group = DispatchGroup()
+            group.enter()
+            self.pendingLookups[package] = group
+            self.pendingLookupsLock.unlock()
+
+            // make sure destination is free.
+            try? self.fileSystem.removeFileTree(packagePath)
+            let isCached = self.cachePath.map{ self.fileSystem.exists($0.appending(packageRelativePath)) } ?? false
+
+            // inform delegate that we are starting to fetch
+            delegateQueue.async {
+                let details = FetchDetails(fromCache: isCached, updatedCache: false)
+                self.delegate?.willFetch(package: package, version: version, fetchDetails: details)
+            }
+
+            let start = DispatchTime.now()
+            self.downloadAndPopulateCache(
+                package: package,
+                version: version,
+                packagePath: packagePath,
+                observabilityScope: observabilityScope,
+                delegateQueue: delegateQueue,
+                callbackQueue: callbackQueue
+            ) { result in
+                // inform delegate that we finished to fetch
+                let duration = start.distance(to: .now())
+                delegateQueue.async {
+                    self.delegate?.didFetch(package: package, version: version, result: result, duration: duration)
+                }
+                // remove the pending lookup
+                self.pendingLookupsLock.lock()
+                self.pendingLookups[package]?.leave()
+                self.pendingLookups[package] = nil
+                self.pendingLookupsLock.unlock()
+                // and done
+                completion(result.map{ _ in packagePath })
+            }
+        }
+    }
+
+    func downloadAndPopulateCache(
+        package: PackageIdentity,
+        version: Version,
+        packagePath: AbsolutePath,
+        observabilityScope: ObservabilityScope,
+        delegateQueue: DispatchQueue,
+        callbackQueue: DispatchQueue,
+        completion: @escaping  (Result<FetchDetails, Error>) -> Void
+    ) {
+        if let cachePath = self.cachePath {
+            do {
+                let relativePath = try package.downloadPath(version: version)
+                let cachedPacakgePath = cachePath.appending(relativePath)
+
+                try self.initializeCacheIfNeeded(cachePath: cachePath)
+                try self.fileSystem.withLock(on: cachedPacakgePath, type: .exclusive) {
+                    // download the package into the cache unless already exists
+                    if try self.fileSystem.validPackageDirectory(cachedPacakgePath) {
+                        // extra validation to defend from racy edge cases?
+                        if self.fileSystem.exists(packagePath) {
```

Comment:
What happens if this was partially downloaded?  We had a case recently where the remedy was to remove the preexisting (possibly partial) directory.  Does that apply here?

Reply:
Oh never mind, I misread the code.  Marking resolved.